From a87989881cbdbf864bccb73a512900623ef08d3b Mon Sep 17 00:00:00 2001
From: Amithash Prasad <amithash@meta.com>
Date: Thu, 4 Sep 2025 20:23:31 -0700
Subject: [PATCH] transport: Improve time validation in
 pldm_transport_send_recv_msg()

There is at least one system where the current value returned by
CLOCK_MONOTONIC is 2589793 which fails the overflow test in
`timeval_is_valid`. Upon close inspecting we dont really want to test
the actual end time but rather the difference (remaining) since the
real risk we are trying to avoid is overflowing when the call to
`timeval_to_msec` is made.

So, moving the checks further down and checking `remaining` instead
of `end`.

Change-Id: If6533855025ff3ba0be19bad791f58ca812b80a7
Signed-off-by: Amithash Prasad <amithash@meta.com>
---
 src/transport/transport.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/src/transport/transport.c b/src/transport/transport.c
index 24c264ade15..7af58a644a3 100644
--- a/src/transport/transport.c
+++ b/src/transport/transport.c
@@ -202,14 +202,14 @@ pldm_transport_send_recv_msg(struct pldm_transport *transport, pldm_tid_t tid,
 	}
 
 	timeradd(&now, &max_response_interval, &end);
-	if (!timeval_is_valid(&end)) {
-		return PLDM_REQUESTER_POLL_FAIL;
-	}
 
 	while (timercmp(&now, &end, <)) {
 		pldm_tid_t src_tid;
 
 		timersub(&end, &now, &remaining);
+		if (!timeval_is_valid(&remaining)) {
+			return PLDM_REQUESTER_POLL_FAIL;
+		}
 
 		/* 0 <= `timeval_to_msec()` <= 4800, and 4800 < INT_MAX */
 		ret = pldm_transport_poll(transport,
-- 
2.47.3

